### PR TITLE
Tzachi indexdby

### PR DESCRIPTION
**Pull request by tzachi-dar**
This PR adds the option to add indexed by to our active android queries.

Here is more info from the commit logs.

SHA-1: 580625f85d51f19658fa4fcdda26f8d3b1b06519

* Use an improved version of the active android library.

The new library is only changed in a few lines (see https://github.com/tzachi-dar/ThreadSafeActiveAndroid) and contains the option to use indexed by.
This allows us to better choose the index that is used for queries, and with this new index we can execute queries faster (see the next checkin as an example).

I have tried Using ANALYZE on the db and this did not have an effect on the indexes that have been used.
The simplest way to check this is to use SQLite browser and run explain query (for example explain query plan SELECT * from BgReadings where sensor = "xx" and  timestamp <= 1000 and  calculated_value = 0 and raw_calculated = 0 ORDER by timestamp DESC)
Then run analyze on the db and run the query again.
As a side note, the analyze command only creates a small table so it is not possible for it to really capture the situation on the DB.

SHA-1: e85cbe8780c6f52824370dc7ebe2d9211ca172a5
* This is an example of using the new indexed by option.

It is not meant to be checked in, but rather to allow people to try it, or see real results.

Please note that result will vary based on your phone, the size of the DB, and the amount of time that the sensor is open (this is because the sensor is used as the default index in this table).

It is printing the time that it takes to execute 10 queries. some queries are executed quite fast, so no gain is possible here.
The "select 8" seems to be improved considerably. The "select 9" seems to be somewhat improved. The "select 7" seems to have no improvement.

Here are *selected* examples:
Select 8 took 3 compared to 13 mily seconds
Select 8 took 2 compared to 10 milli seconds
Select 8 took 1 compared to 10 milly seconds
Select 9 took 6 compared to 11 milly seconds
Select 9 took 65 compared to 138 milly seconds

Select 6 took 1 compared to 6 milly seconds
Select 6 took 1 compared to 7 milly seconds
Select 6 took 1 compared to 7 milly seconds

Select 3 took 57 compared to 102 milly seconds
Select 3 took 40 compared to 57 milly seconds
Select 3 took 34 compared to 38 milly seconds

Please note that although the saving might only be a few ms each time, there can be about 10 queries in the DB at a single second. (See for example what happened on 3-18 14:07:24)

Here is a full log of 5 minutes:
03-18 14:10:00.910 12548 12548 E xxxx    : Select 9 took 60 compared to 135 milly seconds
03-18 14:10:00.984 12548 12548 E xxxx    : Select 7 took 14 compared to 13 milly seconds
03-18 14:10:01.004 12548 12548 E xxxx    : Select 7 took 9 compared to 8 milly seconds
03-18 14:10:01.019 12548 12548 E xxxx    : Select 6 took 1 compared to 9 milly seconds
03-18 14:10:01.039 12548 12548 E xxxx    : Select 7 took 9 compared to 9 milly seconds
03-18 14:10:01.058 12548 12548 E xxxx    : Select 8 took 0 compared to 11 milly seconds
03-18 14:10:18.531 12548 15251 E xxxx    : Select 3 took 57 compared to 102 milly seconds
03-18 14:10:18.633 12548 15251 E xxxx    : Select 3 took 40 compared to 57 milly seconds
03-18 14:10:18.709 12548 15251 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:10:18.786 12548 15251 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:10:18.863 12548 15251 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:10:18.899 12548 15251 E xxxx    : Select 7 took 7 compared to 6 milly seconds
03-18 14:10:18.946 12548 15251 E xxxx    : Select 7 took 12 compared to 11 milly seconds
03-18 14:10:18.984 12548 15251 E xxxx    : Select 7 took 12 compared to 11 milly seconds
03-18 14:10:19.120 12548 15251 E xxxx    : Select 3 took 51 compared to 69 milly seconds
03-18 14:10:19.211 12548 15251 E xxxx    : Select 3 took 37 compared to 48 milly seconds
03-18 14:10:19.287 12548 15251 E xxxx    : Select 3 took 34 compared to 37 milly seconds
03-18 14:10:19.363 12548 15251 E xxxx    : Select 3 took 33 compared to 38 milly seconds
03-18 14:10:19.447 12548 15251 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:10:19.524 12548 15251 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:10:19.599 12548 15251 E xxxx    : Select 3 took 34 compared to 37 milly seconds
03-18 14:10:19.674 12548 15251 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:10:19.752 12548 15251 E xxxx    : Select 3 took 35 compared to 39 milly seconds
03-18 14:10:19.829 12548 15251 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:11:00.928 12548 12548 E xxxx    : Select 9 took 58 compared to 139 milly seconds
03-18 14:11:01.008 12548 12548 E xxxx    : Select 7 took 12 compared to 11 milly seconds
03-18 14:11:01.031 12548 12548 E xxxx    : Select 7 took 11 compared to 10 milly seconds
03-18 14:11:01.043 12548 12548 E xxxx    : Select 6 took 0 compared to 9 milly seconds
03-18 14:11:01.061 12548 12548 E xxxx    : Select 7 took 9 compared to 8 milly seconds
03-18 14:11:01.084 12548 12548 E xxxx    : Select 8 took 1 compared to 12 milly seconds
03-18 14:11:18.617 12548 15335 E xxxx    : Select 3 took 66 compared to 130 milly seconds
03-18 14:11:18.727 12548 15335 E xxxx    : Select 3 took 49 compared to 55 milly seconds
03-18 14:11:18.806 12548 15335 E xxxx    : Select 3 took 34 compared to 40 milly seconds
03-18 14:11:18.887 12548 15335 E xxxx    : Select 3 took 35 compared to 40 milly seconds
03-18 14:11:18.994 12548 15335 E xxxx    : Select 3 took 51 compared to 51 milly seconds
03-18 14:11:19.070 12548 15335 E xxxx    : Select 3 took 34 compared to 37 milly seconds
03-18 14:11:19.148 12548 15335 E xxxx    : Select 3 took 34 compared to 39 milly seconds
03-18 14:11:19.223 12548 15335 E xxxx    : Select 3 took 34 compared to 37 milly seconds
03-18 14:11:19.300 12548 15335 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:11:19.376 12548 15335 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:11:19.453 12548 15335 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:11:19.531 12548 15335 E xxxx    : Select 3 took 34 compared to 39 milly seconds
03-18 14:11:19.608 12548 15335 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:11:19.685 12548 15335 E xxxx    : Select 3 took 35 compared to 38 milly seconds
03-18 14:11:19.764 12548 15335 E xxxx    : Select 3 took 37 compared to 37 milly seconds
03-18 14:12:18.579 12548 15456 E xxxx    : Select 3 took 60 compared to 110 milly seconds
03-18 14:12:18.695 12548 15456 E xxxx    : Select 3 took 48 compared to 57 milly seconds
03-18 14:12:18.774 12548 15456 E xxxx    : Select 3 took 34 compared to 40 milly seconds
03-18 14:12:18.851 12548 15456 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:12:18.929 12548 15456 E xxxx    : Select 3 took 35 compared to 38 milly seconds
03-18 14:12:19.086 12548 15456 E xxxx    : Select 3 took 47 compared to 73 milly seconds
03-18 14:12:19.165 12548 15456 E xxxx    : Select 3 took 34 compared to 40 milly seconds
03-18 14:12:19.242 12548 15456 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:12:19.319 12548 15456 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:12:19.396 12548 15456 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:12:19.473 12548 15456 E xxxx    : Select 3 took 35 compared to 38 milly seconds
03-18 14:12:19.551 12548 15456 E xxxx    : Select 3 took 36 compared to 37 milly seconds
03-18 14:12:19.665 12548 15456 E xxxx    : Select 3 took 53 compared to 45 milly seconds
03-18 14:12:19.763 12548 15456 E xxxx    : Select 3 took 48 compared to 44 milly seconds
03-18 14:12:19.878 12548 15456 E xxxx    : Select 3 took 52 compared to 58 milly seconds
03-18 14:13:18.544 12548 15784 E xxxx    : Select 3 took 67 compared to 116 milly seconds
03-18 14:13:18.652 12548 15784 E xxxx    : Select 3 took 47 compared to 55 milly seconds
03-18 14:13:18.731 12548 15784 E xxxx    : Select 3 took 35 compared to 39 milly seconds
03-18 14:13:18.823 12548 15784 E xxxx    : Select 3 took 35 compared to 41 milly seconds
03-18 14:13:18.901 12548 15784 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:13:18.978 12548 15784 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:13:19.067 12548 15784 E xxxx    : Select 3 took 35 compared to 48 milly seconds
03-18 14:13:19.150 12548 15784 E xxxx    : Select 3 took 34 compared to 37 milly seconds
03-18 14:13:19.227 12548 15784 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:13:19.304 12548 15784 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:13:19.380 12548 15784 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:13:19.458 12548 15784 E xxxx    : Select 3 took 34 compared to 38 milly seconds
03-18 14:13:19.535 12548 15784 E xxxx    : Select 3 took 35 compared to 38 milly seconds
03-18 14:13:19.611 12548 15784 E xxxx    : Select 3 took 35 compared to 37 milly seconds
03-18 14:13:19.687 12548 15784 E xxxx    : Select 3 took 34 compared to 37 milly seconds
03-18 14:13:23.988 12548 15787 E xxxx    : Select 9 took 11 compared to 17 milly seconds
03-18 14:13:24.069 12548 12548 E xxxx    : Select 7 took 54 compared to 50 milly seconds
03-18 14:13:24.105 12548 12548 E xxxx    : Select 7 took 22 compared to 20 milly seconds
03-18 14:13:24.131 12548 12548 E xxxx    : Select 6 took 1 compared to 15 milly seconds
03-18 14:13:24.245 12548 15790 E xxxx    : Select 6 took 1 compared to 10 milly seconds
03-18 14:13:24.297 12548 15790 E xxxx    : Select 6 took 1 compared to 17 milly seconds
03-18 14:14:18.624 12548 15802 E xxxx    : Select 3 took 59 compared to 107 milly seconds
03-18 14:14:18.728 12548 15802 E xxxx    : Select 3 took 41 compared to 56 milly seconds

### CODE REVIEW

**Comment by tolot27** [comment]
I believe you have overseen the ORM and file cache. SQLite does not have a built-in query cache but a database page cache (similar like a file cache with a typical size of 2 MB) and if I understand it correctly, the ActiveAndroid ORM implements an additional table cache. If you execute the same query twice (with same I mean retrieving exactly the same data), you likely get faster results for the second execution because the table data pages were already read by SQLite and also stored in the ORM cache. That explains your "measured" improvements.
If you really want to know which index is used, you have to use the [EXPLAIN QUERY PLAN](https://sqlite.org/eqp.html). I've tested it with one of my larger backups and the `index_BgReadings_timestamp` is always used. Please also read the intention of the [INDEX BY](https://sqlite.org/lang_indexedby.html) clause. It clearly states that it is "not intended for use in tuning the performance of a query."

Also, your example `explain query plan SELECT * from BgReadings where sensor = "xx" and  timestamp <= 1000 and  calculated_value = 0 and raw_calculated = 0 ORDER by timestamp DESC` is likely not representative. Are the where parameters arbitrary chosen or real values? If I execute the query, `index_BgReadings_sensor` is used because the index cardinality is better than `index_BgReadings_timestamp`.

IMHO, there are much easier methods to improve query execution speed. Since almost all of the query retrieving either 1 or a specified number of records, it is simplest to change `index_BgReadings_sensor` into a unique descending index and remove the `ORDER BY timestamp DESC` clause from the queries.
Further optimizations are:
- use the timestamp column as the primary key and omit the `_id` column; this has also the advantage that one of the other where conditions perform an index lookup
- omit the internal row ID (`WITHOUT ROWID`)
- use a subquery to get the last n records: `SELECT * from BgReadings WHERE timestamp = (SELECT MAX(timestamp) FROM BgReadings WHERE timestamp >= 1585486042513 AND timestamp <= 1585878453234 AND calculated_value != 0 AND raw_data != 0);`
- select only fields which are used instead of any (`*`)

It is also possible to use a much faster ORM like ROOM or GreenDAO.

If it is really necessary to optimize the query speed we can collaborate on it.

**Comment by tzachi-dar** [comment]
Thanks for your detailed reply @tolot27. 

TLDR: This PR is short to read, causes no harm to existing code base and allows all further improvments that you have stated.

I have reveresed the order of the queries to remove the caching effect and as you can see some queries still run much faster than before.

Here are some examples:
Select 8 took 18 compared to 1 milly seconds
Select 8 took 18 compared to 3 milly seconds
Select 7 took 10 compared to 1 milly seconds
Select 7 took 10 compared to 2 milly seconds
Select 6 took 8 compared to 1 milly seconds
Select 7 took 14 compared to 4 milly seconds
Select 8 took 23 compared to 2 milly seconds
Select 7 took 11 compared to 1 milly seconds
Select 7 took 9 compared to 0 milly seconds

There are obviously a lot of things that can be done in order to improve the speed of reading from our persitanet database. My proposed change is only a small and simple one and it solves the problem in the cases where the wrong index is used. It is not meant to be the best fix possible (for example, looking at all queries and creating index accordingly will result in better results, but also requires much longer devleopment/test time).

Here is just a few words about indexes in general: The bigger the DB is, scanning it takes more time. Index - where it exists decreases this time to zero. So, if a sensor index is used, the search is being done on the number of objects related to sensor. This limts the search to ~14 days of data (althogh some sensors for example eversense lasts much longer). Doing a query for an hour data (using the timestamp key) will limit the objects that needs to be scanned to an hour, so much more efficent.

In the past there have been many reports about xdrip getting too slow, and it turned out that people have not been stoping and starting sensors. This also caused problems to calibrations, so they can no longer do it with libre or with G6.

My initial motivation was the libreblock table where indexes selection was very wrong, but I believe that if people have to remove objects from the DB there are other issues that we are having.

Indeed you have raised a lot of other optimization options. Do you fill that accepting the change will disturb to doing any of these optimizatons?

**Comment by tzachi-dar** [comment]
Here is a link to the DB so you can try it yourself: https://drive.google.com/file/d/1qXr8AaE6QdxAw_RUQxp0poTWikMzWRgI/view?usp=sharing

As for calling analyze, well I was going to answer that there is no place in xDrip that ever calls Analyze. There is also no such call made from activeandroid. But I was surprised to find out that there is a voice command called database admin that allows one to call analyze. So, trying this right now...

And here are some new results (no change... the indexed queries are faster)
03-26 20:51:33.845  1314  1314 E xxxx    : Select 7 took 12 compared to 0 milly seconds
03-26 20:51:33.858  1314  1314 E xxxx    : Select 6 took 11 compared to 0 milly seconds
03-26 20:51:33.875  1314  1314 E xxxx    : Select 7 took 12 compared to 1 milly seconds
03-26 20:51:33.897  1314  1314 E xxxx    : Select 8 took 15 compared to 0 milly seconds
03-26 20:51:34.007  1314  1314 E xxxx    : Select 8 took 15 compared to 1 milly seconds
03-26 20:51:34.030  1314  1314 E xxxx    : Select 8 took 15 compared to 0 milly seconds
03-26 20:51:34.125  1314  1314 E xxxx    : Select 9 took 41 compared to 42 milly seconds
03-26 20:51:34.166  1314  1314 E xxxx    : Select 7 took 13 compared to 1 milly seconds
03-26 20:51:34.185  1314  1314 E xxxx    : Select 7 took 13 compared to 0 milly seconds
03-26 20:51:34.202  1314  1314 E xxxx    : Select 6 took 12 compared to 2 milly seconds
03-26 20:51:34.225  1314  1314 E xxxx    : Select 7 took 17 compared to 1 milly seconds
03-26 20:51:34.249  1314  1314 E xxxx    : Select 8 took 16 compared to 1 milly seconds
03-26 20:51:38.991  1314  1314 E xxxx    : Select 8 took 15 compared to 1 milly seconds
03-26 20:51:39.015  1314  1314 E xxxx    : Select 8 took 15 compared to 0 milly seconds
03-26 20:51:39.109  1314  1314 E xxxx    : Select 9 took 41 compared to 41 milly seconds
03-26 20:51:39.149  1314  1314 E xxxx    : Select 7 took 12 compared to 0 milly seconds
03-26 20:51:39.167  1314  1314 E xxxx    : Select 7 took 12 compared to 1 milly seconds
03-26 20:51:39.183  1314  1314 E xxxx    : Select 6 took 12 compared to 0 milly seconds

IMHO, the analyze command returns so fast (~few seconds for a DB of 500M) that I don't believe it is actually doing anything useful.
I have only runed analyze once. I'm not sure where connections are being closed but it is also possible to run analyze once every few hours. I can also try this and see if it will help.

**Comment by tolot27** [comment]
> But I was surprised to find out that there is a voice command called "database administration" that allows one to call analyze. 

Ah cool. I did not know it too. @Navid200 that is an important information for you, IMHO.
It is great to see that the queries are now so much faster. That fits to my expectations. 😄

> IMHO, the analyze command returns so fast (~few seconds for a DB of 500M) that I don't believe it is actually doing anything useful.
`ANALYZE` just recreates the index statistics and is very fast. You cannot estimate the index size from the database size. It is much less.

> I have only runed analyze once. I'm not sure where connections are being closed but it is also possible to run analyze once every few hours. I can also try this and see if it will help.

We could run `ANALYZE` with the DailyIntentService.
